### ISSUE TITLE
iw3: --export-depth-only writes constant (flat) frames for all VDA depth models

### ISSUE THREAD

**Issue by brianleake** [nagadomi/nunif#735]
## Summary

With any Video-Depth-Anything model (`VDA_S`, `VDA_B`, `VDA_L`), `--export-disparity --export-depth-only` writes PNG frames that are entirely constant — every pixel the same value, every frame byte-identical. The same command with `Any_V2_L` or `Any_V3_Mono` exports correct per-frame depth.

Normal SBS synthesis with the same VDA models works fine, so the depth reaching the stereo warp is real — only the export path is affected.

## Repro

```
python -m iw3 -i clip.mp4 -o out_dir   --depth-model VDA_S --yes --max-fps 60   --export-disparity --export-depth-only
```

(`clip.mp4` here was 240 frames of 1080p film; any video seems to do.)

## Observed

- All 240 exported PNGs are uint16, constant value ~11822, `std 0.0`, identical md5 across frames.
- Adding `--ema-normalize` changes the constant to 0 (all-black) — still flat.
- Tried: with/without `--scene-detect`, with/without `--depth-aa`, and plain `--export` (RGB+depth) — depth is flat in every combination.
- `Any_V2_L` / `Any_V3_Mono` with identical flags export correct, varying depth.

## Environment

- nunif master (`d23721f`), macOS 26.6.2, Apple Silicon (M4), Python 3.13.13, torch 2.12.1 (MPS available)

Found while benchmarking depth backends for a 2D→3D conversion pipeline — happy to test a fix.

**Comment by brianleake** [comment]
Root cause found (with a workaround). The constant-depth export is not specific to `--export-depth-only` — it is an **MPS allocator-state bug** in the VDA windowed forward, and the export path merely happens to hit the triggering allocation pattern (the export callback batches frames in pairs; `args.batch_size` = 2 there).

Evidence chain, all on torch 2.12.1 / macOS 26.6 / M4:

- Feeding the same 24 frames to `VideoDepthAnythingModel.infer_with_normalize` in groups of 1, 3, 4 or 5 → correct depth. Groups of **2** → every output frame constant. CPU with groups of 2 → correct, so it is MPS-only.
- Hashing the tensor that reaches `VideoDepthAnythingOnline.forward`: **bitwise-identical inputs** (same md5) produce output std 2.72 in a healthy run and std 0.0 in a poisoned one — only the process's prior allocation pattern differs.
- Not input aliasing: `forward(x.clone())` and `forward(x.cpu().to('mps'))` still return the constant.
- Not weight corruption: live GPU `state_dict` matches the on-disk checkpoint bitwise before and after.
- **Workaround**: `torch.mps.empty_cache()` immediately before the window forward restores the correct result exactly (0.000000 → 2.718788, matching the healthy run's value), and subsequent windows stay healthy. Cost is ~ms per 32-frame window.

Local patch (happy to PR if you want it in this shape): wrap `model.forward` at load time in `video_depth_anything_model.py` / `video_depth_anything_streaming_model.py` with an MPS-gated `torch.mps.empty_cache()`. With it, `--export-disparity --export-depth-only` with VDA_S produces correct per-frame depth on Apple Silicon (240/240 frames varying, std ~15k in uint16).

Presumably worth reporting to PyTorch as well — I can file that if useful.

**Comment by brianleake** [comment]
Upstream identified: this is pytorch/pytorch#193487 — "MPS: allocator-state-dependent silently wrong matmul results (10-30% of scale) on 2.7-2.13.0" — same signature (sync doesn't fix, `empty_cache()` does, corruption follows buffer placement after prior allocations). Per that issue, **2.14 nightlies are clean**, so the `empty_cache()` workaround here can be retired once torch ≥ 2.14 is the floor; it is harmless (~ms per 32-frame window) to keep in the meantime.

**Comment by brianleake** [comment]
Two updates on the workaround. (1) On torch 2.12.1 it turns out to be **insufficient for VDA_L** — with the vitl model, 63/240 exported depth frames still came out flat; corruption fires at a site beyond the guarded forward. VDA_S/VDA_B are fully rescued. (2) Verified on `torch 2.15.0.dev20260913`: **unpatched** master runs clean for the exact trigger pattern, VDA_L included, with output bit-identical to healthy runs — confirming pytorch/pytorch#193487 as the root cause and that its fix resolves this completely. So if you take the workaround upstream, gating it on `torch.__version__ < 2.14` (and perhaps skipping vitl on affected versions, or documenting the caveat) would be the accurate shape; it becomes unnecessary once torch ≥ 2.14 is the floor.

**Comment by nagadomi** [comment]
On macOS, PyTorch is not pinned in `requirements-torch.txt`. This is because the MPS implementation always seems to have some issues. If the issues are fixed in 2.14, just update PyTorch to 2.14.
